drivers/bluetooth/hci/hci_nxp_setup.c

```diff
@@ -0,0 +1,1178 @@
+
+#define V3_START_INDICATION 0xabU
+#define V3_HEADER_DATA_REQ  0xa7U
+#define V3_REQUEST_ACK      0x7aU
+#define V3_TIMEOUT_ACK      0x7bU
+#define V3_CRC_ERROR        0x7cU
+
+#define REQ_HEADER_LEN    1U
+#define A6REQ_PAYLOAD_LEN 8U
+#define AbREQ_PAYLOAD_LEN 3U
+
+#define CRC_ERR_BIT          BIT(0)
+#define NAK_REC_BIT          BIT(1)
+#define TIMEOUT_REC_ACK_BIT  BIT(2)
+#define TIMEOUT_REC_HEAD_BIT BIT(3)
+#define TIMEOUT_REC_DATA_BIT BIT(4)
+#define INVALID_CMD_REC_BIT  BIT(5)
+#define WIFI_MIC_FAIL_BIT    BIT(6)
+#define BT_MIC_FAIL_BIT      BIT(7)
+
+#define CMD_HDR_LEN 16
+
+/* CMD5 Header to change bootloader baud rate */
+static uint8_t cmd5_hdrData[CMD_HDR_LEN] = {0x05, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+					    0x2c, 0x00, 0x00, 0x00, 0x77, 0xdb, 0xfd, 0xe0};
+/* CMD7 Header to change timeout of bootloader */
+static uint8_t cmd7_hdrData[CMD_HDR_LEN] = {0x07, 0x00, 0x00, 0x00, 0x70, 0x00, 0x00, 0x00,
+					    0x00, 0x00, 0x00, 0x00, 0x5b, 0x88, 0xf8, 0xba};
+
+enum bt_nxp_ctlr_version {
+	VER1 = 1,
+	VER2,
+	VER3,
+};
+
+struct change_speed_config {
+	uint32_t clk_div_addr;
+	uint32_t clk_div_val;
+	uint32_t uart_clk_div_addr;
+	uint32_t uart_clk_div_val;
+	uint32_t mcr_addr;
+	uint32_t mcr_val;
+	uint32_t reinit_addr;
+	uint32_t reinit_val;
+	uint32_t icr_addr;
+	uint32_t icr_val;
+	uint32_t fcr_addr;
+	uint32_t fcr_val;
+};
+
+#define SEND_BUFFER_MAX_LENGTH  0xFFFFU /* Maximum 2 byte value */
+#define RECV_RING_BUFFER_LENGTH 1024
+
+struct nxp_ctlr_fw_upload_state {
+	uint8_t version;
+	uint8_t hdr_sig;
+
+	uint8_t buffer[A6REQ_PAYLOAD_LEN + REQ_HEADER_LEN + 1];
+
+	uint8_t send_buffer[SEND_BUFFER_MAX_LENGTH + 1];
+
+	struct {
+		uint8_t buffer[RECV_RING_BUFFER_LENGTH];
+		uint32_t head;
+		uint32_t tail;
+		struct k_sem sem;
+	} rx;
+
+	uint16_t length;
+	uint32_t offset;
+	uint16_t error;
+	uint8_t crc8;
+
+	uint32_t last_offset;
+	uint8_t change_speed_buffer[sizeof(struct change_speed_config) + CRC32_LEN];
+
+	uint32_t fw_length;
+	uint32_t current_length;
+	const uint8_t *fw;
+
+	uint32_t cmd7_change_timeout_len;
+	uint32_t change_speed_buffer_len;
+
+	bool wait_hdr_sig;
+
+	bool is_hdr_data;
+	bool is_error_case;
+	bool is_cmd7_req;
+	bool is_entry_point_req;
+
+	uint8_t last_5bytes_buffer[6];
+};
+
+static struct nxp_ctlr_fw_upload_state fw_upload;
+
+static int fw_upload_read_data(uint8_t *buffer, uint32_t len)
+{
+	while (len > 0) {
+		k_sem_take(&fw_upload.rx.sem, K_FOREVER);
+		*buffer = fw_upload.rx.buffer[fw_upload.rx.tail];
+		buffer++;
+		fw_upload.rx.tail++;
+		fw_upload.rx.tail = fw_upload.rx.tail % sizeof(fw_upload.rx.buffer);
+		len--;
+	}
+	return 0;
+}
+
+static void fw_upload_read_to_clear(void)
+{
+	uint32_t key;
+
+	key = irq_lock();
+	k_sem_reset(&fw_upload.rx.sem);
+	fw_upload.rx.head = 0;
+	fw_upload.rx.tail = 0;
+	irq_unlock(key);
+}
+
+static int fw_upload_wait_for_hdr_sig(void)
+{
+	int err;
+	int64_t end;
+	char c;
+
+	end = k_uptime_get() + CONFIG_BT_NXP_CTLR_WAIT_HDR_SIG_TIMEOUT;
```

Comment:
Use `k_timepoint_t`

Reply:
It seems `k_timepoint_t` depends on CONFIG_SYS_CLOCK_EXISTS. If CONFIG_SYS_CLOCK_EXISTS cannot be enabled, the feature cannot be supported.
It is a limitation for this PR.

For k_uptime_get(), there is not such limitation.